kv/kvserver: TestReplicaStateMachineChangeReplicas failed

kv/kvserver.TestReplicaStateMachineChangeReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3338274&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3338274&tab=artifacts#/) on master @ [61bd543ba7288c8f0eed6cddded7b219c9d1fcd4](https://github.com/cockroachdb/cockroach/commits/61bd543ba7288c8f0eed6cddded7b219c9d1fcd4):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x19a7299]

```
Stack: 

```
goroutine 11681 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000826e00, 0x5ce2578, 0xc000b85530)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x126
panic(0x41d31c0, 0x7bb0a30)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).getBreaker(0xc0011043c0, 0x7f0100000002, 0xc00074d380)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:236 +0xb9
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).GetCircuitBreaker(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:229
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).SendAsync(0xc0014a01c0, 0xc000d3c540, 0x1, 0x499100)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/raft_transport.go:553 +0xf4
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRequest(0xc000254000, 0x5ce2578, 0xc000ff9e90, 0xc000d3c540, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:1335 +0x7c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessage(0xc000254000, 0x5ce2578, 0xc000ff9e90, 0x3, 0x2, 0x1, 0x6, 0x6, 0xc, 0xc001012db0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:1306 +0x49b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessages(0xc000254000, 0x5ce2578, 0xc000ff9e90, 0xc000c1d8c0, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:1244 +0x1c5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(0xc000254000, 0x5ce2578, 0xc000ff9e90, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:724 +0x84a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(0xc000254000, 0x5ce2578, 0xc000ff9e90, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:487 +0x11c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc0003c0000, 0x5ce2578, 0xc000b85530, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:515 +0x145
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker(0xc0001ae280, 0x5ce2578, 0xc000b85530)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:284 +0x2c2
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc000826e00, 0x5ce2578, 0xc000b85530, 0x0, 0x0, 0xc0008258e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=true/deprecated=false
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestReplicaStateMachineChangeReplicas PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaStateMachineChangeReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_application_state_machine_test.go
+++ b/pkg/kv/kvserver/replica_application_state_machine_test.go
@@ -77,7 +77,7 @@ func TestReplicaStateMachineChangeReplicas(t *testing.T) {
 
 				confChange = raftpb.ConfChange{
 					Type:   raftpb.ConfChangeAddNode,
-					NodeID: uint64(addedReplDesc.NodeID),
+					NodeID: uint64(addedReplDesc.ReplicaID),
 				}
 			} else {
 				// Remove ourselves from the Range.
@@ -100,7 +100,7 @@ func TestReplicaStateMachineChangeReplicas(t *testing.T) {
 
 				confChange = raftpb.ConfChange{
 					Type:   raftpb.ConfChangeRemoveNode,
-					NodeID: uint64(removedReplDesc.NodeID),
+					NodeID: uint64(removedReplDesc.ReplicaID),
 				}
 			}
 

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -49,6 +49,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/txnwait"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/rpc"
+	"github.com/cockroachdb/cockroach/pkg/rpc/nodedialer"
 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
@@ -218,19 +219,24 @@ func (tc *testContext) StartWithStoreConfigAndVersion(
 	t testing.TB, stopper *stop.Stopper, cfg StoreConfig, bootstrapVersion roachpb.Version,
 ) {
 	tc.TB = t
+	ctx := context.Background()
 	// Setup fake zone config handler.
 	config.TestingSetupZoneConfigHook(stopper)
+	rpcContext := rpc.NewContext(rpc.ContextOptions{
+		TenantID:   roachpb.SystemTenantID,
+		AmbientCtx: cfg.AmbientCtx,
+		Config:     &base.Config{Insecure: true},
+		Clock:      cfg.Clock,
+		Stopper:    stopper,
+		Settings:   cfg.Settings,
+	})
+	grpcServer := rpc.NewServer(rpcContext) // never started
 	if tc.gossip == nil {
-		rpcContext := rpc.NewContext(rpc.ContextOptions{
-			TenantID:   roachpb.SystemTenantID,
-			AmbientCtx: cfg.AmbientCtx,
-			Config:     &base.Config{Insecure: true},
-			Clock:      cfg.Clock,
-			Stopper:    stopper,
-			Settings:   cfg.Settings,
-		})
-		server := rpc.NewServer(rpcContext) // never started
-		tc.gossip = gossip.NewTest(1, rpcContext, server, stopper, metric.NewRegistry(), zonepb.DefaultZoneConfigRef())
+		tc.gossip = gossip.NewTest(1, rpcContext, grpcServer, stopper, metric.NewRegistry(), zonepb.DefaultZoneConfigRef())
+	}
+	if tc.transport == nil {
+		dialer := nodedialer.New(rpcContext, gossip.AddressResolver(tc.gossip))
+		tc.transport = NewRaftTransport(cfg.AmbientCtx, cfg.Settings, dialer, grpcServer, stopper)
 	}
 	if tc.engine == nil {
 		disableSeparatedIntents :=
@@ -249,10 +255,6 @@ func (tc *testContext) StartWithStoreConfigAndVersion(
 		}
 		stopper.AddCloser(tc.engine)
 	}
-	if tc.transport == nil {
-		tc.transport = NewDummyRaftTransport(cfg.Settings)
-	}
-	ctx := context.Background()
 	if tc.store == nil {
 		cv := clusterversion.ClusterVersion{Version: bootstrapVersion}
 		cfg.Gossip = tc.gossip

